Implement Value Caching and Diffing for Optimized Change Detection

faizanu94 opened the pull request:
This pull request introduces value caching and diffing mechanisms to optimize change detection. By caching previous values and diffing them against current values, we ensure that only the property that has changed is passed to the callback function as discussed in #1

Changes
- Added a value caching system to store previous state values
- Implemented a diffing algorithm to efficiently compare current values against cached values
- Modified callback functions to receive only the properties that have changed, based on the diffing results

faizanu94 committed:
Implement Value Caching and Diffing for Optimized Change Detection

bramus commented:
@faizanu94 I’m not sure if you are looking for a review yet or not, but here goes: the core of this PR looks really good.

The only thing that is missing is a flag to get back the old behavior, where an author can choose between getting only these `changedProperties` (default behavior) or getting back all the monitored properties.

The logic for that is similar to the flag you wrote in #1.

_(Once this PR is in, we can pick up #1 again to let users choose between the current simple format, or the more rich format)_

faizanu94 committed:
add toggle for observability mode

faizanu94 committed:
Update README.md

faizanu94 commented:
Thanks for the reminder about taking care of the old behavior - I’ve added the flag `observeAllProperties` to toggle between the old and new behaviors, and updated the README accordingly. All set now!

bramus commented:
> I’ve added the flag `observeAllProperties` to toggle between the old and new behaviors, and updated the README accordingly. All set now!

Instead of a `boolean` here, this needs to be the `{ callbackMode: CallbackMode.INDIVIDUAL }` that you introduced in #3. Maybe a better name would be `NotificationMode`.

faizanu94 committed:
added notification mode

faizanu94 committed:
Update README.md

faizanu94 commented:
Thanks for the feedback! I've replaced the boolean flag with `NotificationMode` as suggested, using `NotificationMode.INDIVIDUAL` and `NotificationMode.ALL`. This approach can easily scale to other options like `ReturnFormat` in the future. README has been updated accordingly

bramus commented on README.md:
```diff
@@ -50,6 +54,10 @@ cssStyleObserver.attach(document.body);  /* Attach observer to `document.body` *
 
+### Configuration options
+
+* notificationMode (NotificationMode, default: INDIVIDUAL): Determines whether to observe and return all properties (ALL) or only the changed ones (INDIVIDUAL)
```
```suggestion
* `notificationMode` (`NotificationMode`, default: `CHANGED_ONLY`): Determines whether to pass all properties (`ALL`) or only the changed ones (`CHANGED_ONLY`) into the callback
```

bramus commented on README.md:
```diff
@@ -40,7 +40,11 @@ const cssStyleObserver = new CSSStyleObserver(
+    },                                                 
+    /* Configuration options */
+    {
+      notificationMode?: NotificationMode.INDIVIDUAL
```
```suggestion
      notificationMode?: NotificationMode.CHANGED_ONLY
```

bramus commented on src/index.ts:
```diff
@@ -19,6 +19,21 @@ export type CSSStyleObserverCallback = (
+/**
+ * Enum for callback modes
+ */
+export enum NotificationMode {
+  INDIVIDUAL = 'individual',
+  ALL = 'all',
+}
```
```suggestion
export enum NotificationMode {
  CHANGED_ONLY = 'changed_only',
  ALL = 'all',
}
```

bramus commented on src/index.ts:
```diff
@@ -37,14 +52,18 @@ export class CSSStyleObserver {
     this._callback = callback;
     this._targetElement = null;
+    this._cachedValues = {};
+    this._notificationMode = options.notificationMode ?? NotificationMode.INDIVIDUAL;
```
```suggestion
    this._notificationMode = options.notificationMode ?? NotificationMode.CHANGED_ONLY;
```

bramus reviewed (changes requested):
This is looking great! In hindsight the `INDIVIDUAL` mode is incorrectly named, as it can report two changed properties at once.

I have made suggestions (which you can apply in the GitHub UI by simply clicking the button) to rename it to `CHANGED_ONLY`.

Also reworded the text in the README.

faizanu94 committed:
Update README.md

Co-authored-by: Bramus <bramus@bram.us>

faizanu94 committed:
Update README.md

Co-authored-by: Bramus <bramus@bram.us>

faizanu94 committed:
Update src/index.ts

Co-authored-by: Bramus <bramus@bram.us>

faizanu94 committed:
Update src/index.ts

Co-authored-by: Bramus <bramus@bram.us>

faizanu94 commented:
Thanks for the suggestions - I’ve applied the changes. Appreciate the guidance!
